src/modules/adaptors/Morpho/MorphoBlue/MorphoBlueCollateralAdaptor.sol

```diff
@@ -0,0 +1,224 @@
+    /**
+     * @notice User withdraws are NOT allowed from this position.
+     * NOTE: collateral withdrawal calls directly from users disallowed for now.
+     */
+    function withdraw(uint256, address, bytes memory, bytes memory) public pure override {
+        revert BaseAdaptor__UserWithdrawsNotAllowed();
+    }
+
+    /**
+     * @notice This position is a debt position, and user withdraws are not allowed so
+     *         this position must return 0 for withdrawableFrom.
+     * NOTE: collateral withdrawal calls directly from users disallowed for now.
+     */
+    function withdrawableFrom(bytes memory, bytes memory) public pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Returns the cellar's balance of the collateralAsset position in corresponding Morpho Blue market.
+     * @param adaptorData containing the abi encoded Morpho Blue market Id.
+     * @return Cellar's balance of provided collateral to specified MB market.
+     * @dev normal static call, thus msg.sender for most-likely Sommelier usecase is the calling cellar.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        Id id = abi.decode(adaptorData, (Id));
+        return _userCollateralBalance(id, msg.sender);
+    }
+
+    /**
+     * @notice Returns collateral asset.
+     * @param adaptorData containing the abi encoded Morpho Blue market Id.
+     * @return The collateral asset in ERC20 type.
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        Id id = abi.decode(adaptorData, (Id));
+        MarketParams memory market = morphoBlue.idToMarketParams(id);
+        return ERC20(market.collateralToken);
+    }
+
+    /**
+     * @notice This adaptor returns collateral, and not debt.
+     * @return Whether or not this position is a debt position
+     */
+    function isDebt() public pure override returns (bool) {
+        return false;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    /**
+     * @notice Allows strategists to add collateral to the respective cellar position on specified MB Market, enabling borrowing.
+     * @param _id identifier of a Morpho Blue market.
+     * @param _collateralToDeposit The amount of `collateralToken` to add to specified MB market position.
+     */
+    function addCollateral(Id _id, uint256 _collateralToDeposit) public {
+        _validateMBMarket(_id);
+        MarketParams memory market = morphoBlue.idToMarketParams(_id);
+        ERC20 collateralToken = ERC20(market.collateralToken);
+        uint256 amountToDeposit = _maxAvailable(collateralToken, _collateralToDeposit);
+        collateralToken.safeApprove(address(morphoBlue), amountToDeposit);
+        _addCollateral(market, amountToDeposit);
+        // Zero out approvals if necessary.
+        _revokeExternalApproval(collateralToken, address(morphoBlue));
```

Comment:
This logic is very similar to `deposit` aside from the incoming transfer from `msg.sender`. Would it be interesting to factorize this logic?

Reply:
Good catch, thanks for this, yes factorizing is interesting. I've implemented some factorizing. 

NOTE: it can be further refactored if we use `marketParams` as an encoded adaptorData param. I will continue looking into it.